tensorflow/training/docker/2.9/py3/cu112/Dockerfile.gpu

```diff
@@ -0,0 +1,287 @@
+  make install && \
+  ldconfig && \
+  cd / && \
+  rm -rf /tmp/openmpi
+
+###########################################################################
+# Horovod & its dependencies
+###########################################################################
+
+# Create a wrapper for OpenMPI to allow running as root by default
+RUN mv $OPEN_MPI_PATH/bin/mpirun $OPEN_MPI_PATH/bin/mpirun.real \
+ && echo '#!/bin/bash' > $OPEN_MPI_PATH/bin/mpirun \
+ && echo 'mpirun.real --allow-run-as-root "$@"' >> $OPEN_MPI_PATH/bin/mpirun \
+ && chmod a+x $OPEN_MPI_PATH/bin/mpirun
+
+# Configure OpenMPI to run good defaults:
+#   --bind-to none --map-by slot --mca btl_tcp_if_exclude lo,docker0
+RUN echo "hwloc_base_binding_policy = none" >> $OPEN_MPI_PATH/etc/openmpi-mca-params.conf \
+ && echo "rmaps_base_mapping_policy = slot" >> $OPEN_MPI_PATH/etc/openmpi-mca-params.conf
+
+# Set default NCCL parameters
+RUN echo NCCL_DEBUG=INFO >> /etc/nccl.conf
+ENV LD_LIBRARY_PATH=$OPEN_MPI_PATH/lib/:$EFA_PATH/lib/:$LD_LIBRARY_PATH
+# /usr/local/lib/libpython* needs to be accessible for dynamic linking
+ENV LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH
+ENV PATH=$OPEN_MPI_PATH/bin/:$PATH
+ENV PATH=$OPEN_MPI_PATH/nvidia/bin:$PATH
+
+# SSH login fix. Otherwise user is kicked off after login
+RUN mkdir -p /var/run/sshd \
+   && sed 's@session\s*required\s*pam_loginuid.so@session optional pam_loginuid.so@g' -i /etc/pam.d/sshd
+
+# Create SSH key.
+RUN mkdir -p /root/.ssh/ \
+   && ssh-keygen -q -t rsa -N '' -f /root/.ssh/id_rsa \
+   && cp /root/.ssh/id_rsa.pub /root/.ssh/authorized_keys \
+   && printf "Host *\n  StrictHostKeyChecking no\n" >> /root/.ssh/config
+
+WORKDIR /
+
+RUN apt-get update \
+   && apt-get install -y --no-install-recommends \
+   libbz2-dev \
+   libc6-dev \
+   libffi-dev \
+   libgdbm-dev \
+   libncursesw5-dev \
+   libreadline-gplv2-dev \
+   libsqlite3-dev \
+   libssl-dev \
+   tk-dev \
+   ffmpeg \
+   libsm6 \
+   libxext6 \
+   # remove libsasl2-2 after Nvidia docker upstream fix CVE-2022-24407
+   && rm -rf /var/lib/apt/lists/* \
+   && apt-get clean
+
+RUN wget https://www.python.org/ftp/python/$PYTHON_VERSION/Python-$PYTHON_VERSION.tgz \
+   && tar -xvf Python-$PYTHON_VERSION.tgz \
+   && cd Python-$PYTHON_VERSION \
+   && ./configure --enable-shared \
+   && make -j $(nproc) \
+   && make install \
+   && rm -rf ../Python-$PYTHON_VERSION* \
+   # Starting from Python39, a xxx.pem file will be generated under /tmp folder during installation. Remove it to complete cleanup after installation from python source.
+   && rm /tmp/*.pem
+
+RUN ${PIP} --no-cache-dir install --upgrade \
+   pip \
+   setuptools
+
+# Some TF tools expect a "python" binary
+RUN ln -s $(which ${PYTHON}) /usr/local/bin/python \
+   && ln -s $(which ${PIP}) /usr/bin/pip
+
+# install PyYAML==5.1.2 to avoid conflict with latest awscli
+# # python-dateutil==2.8.0 to satisfy botocore associated with latest awscli
+RUN ${PIP} install --no-cache-dir -U \
```

Comment:
Are we keeping the pinned versions of pip package as of TF2.8? Any modification required?

Reply:
Updated